plutus-contract-exe/src/Language/Plutus/Contract.hs

```diff
@@ -1,163 +1,38 @@
-after running 'lock' endpoint. So the contract is finished for the initiator of
-the game. Participants in the game may still submit guesses etc.
-
-('ContractFinished' seems to be tied closely to contract roles, and we
-should think about this some more. This is the difference between the global
-(ie. blockchain/utxo set) view and the local (what can that particular role do)
-view of a contract.)
-
--}
-
-
--- | Events that are produced by contract endpoints.
-data ContractOut =
-      SubmitTransaction UnbalancedTx
-      -- ^ Produce an 'UnbalancedTx'. See note [Unbalanced transactions]
-
-      | StartWatching L.Address
-      -- ^ Start watching an address. This adds the address to the set of
-      -- "interesting addresses" of this contract instance.
-
-      | ContractError Text
-      -- ^ An error occurred during contract execution.
-      --   NOTE: Should we also set the appropriate HTTP status code?
-
-      | ContractFinished
-      -- ^ Execution of the contract has ended. No further ledger updates are
-      --   required and no user actions are possible. All triggers associated
-      --   with this contract instance can be deleted. See note
-      --   [ContractFinished event]
+import           Control.Monad                        ((>=>))
+import           Data.Aeson                           (FromJSON)
+import qualified Data.Aeson                           as Aeson
 
-      deriving stock (Eq, Ord, Show, Generic)
-      deriving anyclass (Aeson.FromJSON, Aeson.ToJSON)
+import           Language.Plutus.Contract.Contract    as Contract
+import           Language.Plutus.Contract.Event       as Event
+import           Language.Plutus.Contract.Transaction as Transaction
 
--- | Events that inform the contract about changes to the ledger state.
-data LedgerUpdate =
-    OutputSpent L.TxOutRef L.Tx
-    -- ^ An output from one of the interesting addresses was spent. Includes
-    --   the transaction that spent the output.
-    --   The transaction *spends* the UTXO referenced by 'L.TxOutRef'.
+import           Ledger.Tx                            (Address, Tx)
 
-    | OutputAdded L.TxOutRef L.Tx
-    -- ^ An output was produced to one of the interesting addresses.
-    --   The transaction *produces* the UTXO referenced by 'L.TxOutRef'
+type PlutusContract a = Contract Step Event a
 
-    -- NOTE: A transaction that spends an output from an interesting address and
-    --       produces a new output to that address will trigger two events:
-    --       'OutputSpent' and 'OutputAdded' (with the same 'L.Tx'
-    --       argument but with different 'L.TxOutRef' arguments).
-    --
-    --       TODO: Does it make sense to have a 3rd event type for this
-    --       situation?
+-- | Watch an 'Address', returning the next transaction that changes it
+watchAddress :: Address -> PlutusContract Tx
+watchAddress a = await (Event.addr a) (fmap snd <$> ledgerUpdate) -- TODO: Check that tx actually involves the address
```

Comment:
This will return the next transaction that changes *any* address we're watching, right?

Reply:
Fixed